Add reflect `TypeLambda.paramVariances`

nicolasstucki opened the pull request:
Fixes #16734

nicolasstucki commented:
@neko-kai could you check if the test example is representative of your use cases?

neko-kai commented:
@nicolasstucki Almost, there's the type lambda case `variances[[A] =>> KCov[A]]` - I think that's the case that broke in 3.3.0 - before that declaredVariances return +A for this type lambda, now it doesn't - so we'd need a function that tells the real computed variance of a A inside this type lambda - even if type lambda itself is not variance-annotated. Thank you for helping us here btw!

neko-kai commented:
@nicolasstucki Almost, there's the type lambda case `variances[[A] =>> KCov[A]]` - I think that's the case that broke in 3.3.0 - before that declaredVariances return +A for this type lambda, now it doesn't - so we'd need a function that tells the real computed variance of a A inside this type lambda - even if type lambda itself is not variance-annotated. Thank you for helping us here btw!

There's also the type param case

```scala
variances[[A, F[+_]] =>> F[A]]
```

nicolasstucki commented on tests/run-macros/i16734/Test_2.scala:
```diff
@@ -0,0 +1,35 @@
+  println(variances[KFunky])
+  println(variances[[A, F[B]] =>> F[A]])
+  println(variances[[A, F[+B]] =>> F[A]])
+  println(variances[[A, F[-B]] =>> F[A]])
```
@neko-kai to make these work, I used `paramVariance` instead of `declaredVariances`. Is this the functionality you need? See the variances in the check file. If it is what you need then you could also access this method reflectively as a workaround in `3.3.0`.

nicolasstucki commented:
> Currently the result for `A` is that it is still "declared" invariant, which is true, but would it be also possible to expose how the compiler computes the real variance of `A`? (which is what #16734 is asking for) or should macro authors compute that themselves?

I changed that when I switched to `paramVariances`.

smarter commented on library/src/scala/quoted/Quotes.scala:
```diff
@@ -3235,8 +3235,16 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
+         *  Variance flags can be one of `Flags.{Covariant, Contravariant, EmptyFlags}`.
+         */
+        @experimental
+        def paramVariances: List[Flags]
```
IMO, `declaredVariances` is an implementation detail, although it is pickled in tasty.

sjrd replied:
I agree with @smarter. `declaredVariances` is an implementation detail of the compiler/the storage, and not a semantic information. It should not be exposed.

smarter commented on tests/run-custom-args/tasty-inspector/stdlibExperimentalDefinitions.scala:
```diff
@@ -66,6 +66,7 @@ val experimentalDefinitionInLibrary = Set(
   // Should be stabilized in 3.4.0
   "scala.quoted.Quotes.reflectModule.defnModule.FunctionClass",
   "scala.quoted.Quotes.reflectModule.FlagsModule.AbsOverride",
+  "scala.quoted.Quotes.reflectModule.TypeLambdaMethods.declaredVariances",
```
```suggestion
  "scala.quoted.Quotes.reflectModule.TypeLambdaMethods.paramVariances",
```

smarter commented on compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:
```diff
@@ -2212,6 +2212,8 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
       extension (self: TypeLambda)
         def param(idx: Int): TypeRepr = self.newParamRef(idx)
         def paramBounds: List[TypeBounds] = self.paramInfos
+        def paramVariances: List[Flags] =
```
Having this just on `TypeLambda` is a bit limiting since classes also have parameter with variance annotations, it would be nice if the same API worked for TypeLambda, TypeRef of classes and TypeRef of parameterized type aliases.

nicolasstucki replied:
I added `Symbol.paramVariance` and added tests for all these cases. I did not add it on `TypeRef` because it is more consistent if we go to the definition of the class and ask for the variances on it.

sjrd commented on library/src/scala/quoted/Quotes.scala:
```diff
@@ -3235,8 +3235,16 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
     /** Extension methods of `TypeLambda` */
     trait TypeLambdaMethods:
       extension (self: TypeLambda)
+        /** Reference to the i-th parameter */
+        /** Type bounds of the i-th parameter */
         def param(idx: Int) : TypeRepr
         def paramBounds: List[TypeBounds]
```
```suggestion
        /** Reference to the i-th parameter */
        def param(idx: Int) : TypeRepr
        /** Type bounds of the i-th parameter */
        def paramBounds: List[TypeBounds]
```

nicolasstucki commented on library/src/scala/quoted/Quotes.scala:
```diff
@@ -4055,8 +4063,17 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
+        /** Variance flags for of this type parameter.
+         *
+         *  Variance flags can be one of `Flags.{Covariant, Contravariant, EmptyFlags}`.
+         *  If this is not the symbol of a type parameter the result is `Flags.EmptyFlags`.
```
We do not show the distinction between type and term flags in the reflection API.

bishabosha commented:
needs a rebase
